File: eunbi/week13/할인행사.java
package algorithmStudy;

import java.util.*;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        Map<String, Integer> wantMap = new HashMap<String, Integer>();
        
        for(int i=0; i<want.length; i++)
            wantMap.put(want[i], number[i]);
        
        for(int j=0; j<discount.length - 9; j++){                 
            int idx = 0;
            Map<String, Integer> map = new HashMap<String, Integer>();
            
            for(int i=0; i<10; i++)
                map.put(discount[i+j], map.getOrDefault(discount[i+j], 0) + 1);
            
            for (Map.Entry<String, Integer> entry : wantMap.entrySet()) 
                if(map.containsKey(entry.getKey()) && entry.getValue()<=map.get(entry.getKey()))
                    idx++;
            
            if(idx == want.length)
                answer++;
        }

        return answer;
    }
}

File: SiHyeon/week13/할인 행사.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<string> want, vector<int> number, vector<string> discount)
{
    int answer = 0;
    // count(시작, 시작+10, 찾는 원소)로 계산
    for (int day = 0; day <= discount.size() - 10; day++)
    {
        int match = 0; // 살 수 있는 want의 수
        for (int i = 0; i < want.size(); i++)
        {
            if (count(discount.begin() + day, discount.begin() + day + 10, want[i]) >= number[i])
            {
                match++;
            }
            else
                break;
        }
        if (match == want.size()) // want한 모든걸 살수있다면
        {
            answer++;
        }
    }
    return answer;
}